### ISSUE TITLE
Full stack HTML Accessibility Lab: Build a checkout page - exercise not in line with good ARIA practice

### ISSUE THREAD

**Issue by RenateRoke** [freeCodeCamp/freeCodeCamp#59725]
In the accessibility lab (Build a checkout page) of the full stack curriculum HTML course (https://www.freecodecamp.org/learn/full-stack-developer/lab-checkout-page/build-a-checkout-page), students are tought to put the aria-required attribute on two input elements, which also have a regular HTML required attribute. This is not in line with how ARIA is supposed to be used. 

It teaches new developers that any ARIA use is better than no ARIA use, which is setting them up for a lot of accessibility issues in their future work. (This seems to teach that you should always add WAI-ARIA eventhough it doesn't add anything, which is setting up for accessibility errors in the future. The webaim million (https://webaim.org/projects/million/) shows that webpages that use ARIA contain twice the amount of automatically found accessibility errors than pages without ARIA).

As you know, the first rule of WAI-ARIA is to always prefer native HTML over WAI-ARIA if applicable. Since the HTML required attribute is already used, aria-required is redundant since the required attribute is well supported in browsers and screenreaders today. 

I can image wanting to teach ARIA in the HTML course (without any JavaScript) can be a challenge, but there are better options than the current lab, for example:

- Identifying multiple nav-elements on a page with aria-label or aria-labelledby, on a page where there are multiple menu's with different purposes (main menu, sidebar menu etc)
- Hiding purely decorative icons with aria-hidden, for example, in a set of social media links where the name of the platform is already present in the linktext.
- Using aria-describedby to provide helpful suggestions for input elements (which doesn't require a whole new Lab setup).

**Comment by RenateRoke** [comment]
@huyenltnguyen the lab asks for both required and aria-required, so either removing aria-required and leaving it at that works, or something like aria-describedby to programmatically link helpful instructions below inputs works as well to improve it.

**Comment by ilenia-magoni** [comment]
cc/ @bbsmooth can you take a look please?

**Comment by bbsmooth** [comment]
Maybe I am missing something here, but I only see the instructions asking for `aria-required`, not both `aria-required` and `required`. Regardless, you definitely don't need both any more and it is probably best to use `required` over `aria-required`.

**Comment by bbsmooth** [comment]
@huyenltnguyen I like your idea of the `aria-hidden` asterisk. I always recommend this to clients.

**Comment by RenateRoke** [comment]
@bbsmooth see check 11: 11. At least two of your input elements should be required and have an aria-required attribute set to true.

**Comment by ilenia-magoni** [comment]
@bbsmooth it wants both of them

https://github.com/freeCodeCamp/freeCodeCamp/blob/6d2bd791458a23a53ad7701662cf49fc462d1467/curriculum/challenges/english/25-front-end-development/lab-checkout-page/66da326c02141df538f29ba5.md#L101-L108

**Comment by jdwilkin4** [comment]
I have read through the whole conversation and here is how we should move forward based on the conversation.

- update the tests, hints and user story to just have them use `required` instead of `aria-required`
- add a user story, hints and tests to ask campers to add an asterisk next to the input labels to indicate that the field is required. Also ask them to use `aria-hidden` here to hide the asterisk from non-sighted users.

I am opening this up for `help-wanted` again

**Comment by jdwilkin4** [comment]
> For (2), [the lectures before this lab](https://www.freecodecamp.org/learn/full-stack-developer/#lecture-importance-of-accessibility-and-good-html-structure) cover a handful of ARIA attributes, so I think the lab could be made a little more comprehensive to include those.

@huyenltnguyen 

There are two open PRs for new HTML accessibility labs, here

https://github.com/freeCodeCamp/freeCodeCamp/issues?q=is%3Apr+is%3Aopen+author%3Alarymak

So it is best to look at those to see if any more ARIA attributes should be covered there 👍🏾

**Comment by RenateRoke** [comment]
> > At least two of your input elements should be required and have an aria-required attribute set to true.
> 
> Ya, I saw that, I just didn't take it to mean that it was requiring the `required` attribute. I took it to mean that the element should be marked as required using the `aria-required` attribute. To me, the phrase "elements should be required" does not explicitly mean "use the `required` attribute". There are several ways to mark an element as required, include the `required` attribute and the `aria-required` attribute. So "elements should be required" can mean either way and the instructions are telling you to use `aria-required` which means you don't need to use `required`. But I'm a pretty literal person, so my reading of it may not the be same as everyone else's.

Oh right, ofcourse, sorry! No the test fails without both required and aria-required, I should have been more clear.

**Comment by RenateRoke** [comment]
@jdwilkin4 Since it's a WCAG for success criteria 3.3.2 Labels or instructions best practice that if you use the * to indicate required fields for sighted users, you also include an explanatory text like 'Fields with * are required', only hiding the * will now leave screenreader users with a confusing piece of text. Another option would be to hide thát text as well ofcourse, but it's still not a great method to teach what ARIA can actually do. My advice would be to go for the aria-describedby option to programmatically include helpful suggestions to inputs, to teach accessibility best practices.

**Comment by bbsmooth** [comment]
> only hiding the * will now leave screenreaderusers with a confusing piece of text.

Can you expand on this. I don't understand what you mean.

**Comment by RenateRoke** [comment]
@bbsmooth If we were to add the text 'Fields with * are required' above our form (a best practice if using * instead of adding 'required' or 'optional' to our labels), and then hide the *'s in our labels or descriptions, it can be a bit confusing for screenreader users. It's not a big issue ofcourse, but maybe not the best introduction to accessibility for new developers (in my opinion).  You coúld hide that text as well, but we should be cautious with hiding content from people. 

I would love to see an accessibility lab where the inclusion of ARIA ádds something to the user experience for several groups. Teaching aria-describedby helps screenreader users, but also encourages using text instead of placeholders for helpful text, which helps users with cognitive disabilities as well (since placeholders disappear). 

You already have provided the information to students with this exact example scenario in this lecture: https://www.freecodecamp.org/learn/full-stack-developer/lecture-importance-of-accessibility-and-good-html-structure/what-is-the-aria-describedby-attribute. Which could also be a great place to add some information about browse and focus mode and how adding text in between formfields could be missed by blind screenreader users.  Just my two cents ofcourse, the information in the course is already so much better than what a lot of education material provides!

**Comment by bbsmooth** [comment]
> then hide the *'s in our labels or descriptions, it can be a bit confusing for screenreader users

I'm not sure I agree. The input will have the required attribute, so screen reader users will hear that the field is required. Not hearing "star" because we are hiding the asterisk from them isn't going to cause any confusion. I think screen reader users understand the asterisk is for sighted people. And I think it can easily be argued that hearing "star" is just adding needless duplication. Being accessible is about providing a good experience for people. It doesn't mean it literally has to be exactly the same experience. Sometimes less is better.

But I do agree with you that aria-describedby would be a great one to use as well.

**Comment by RenateRoke** [comment]
Hi @xzq050412, great to hear you want to learn accessibility! I advise you to read about aria and what it's for, and especially when you shouldn't use it. 

Axe is great place to start automated testing, but don't forget automated testing doesn't cover a lot (yet). Try to get some practice in with manual testing if you can :D

**Comment by jdwilkin4** [comment]
@hmloftus 

Issue is still open.

Based on the discussion, here is the consensus

- update the tests, hints and user story to just have them use required instead of aria-required

- add a user story, hints and tests to ask campers to add an asterisk next to the input labels to indicate that the field is required. Also ask them to use aria-hidden here to hide the asterisk from non-sighted users.

- add user story for aria-describedby

**Comment by hmloftus** [comment]
I have opened a PR for this issue here: https://github.com/freeCodeCamp/freeCodeCamp/pull/60447
